File: examples/spinwave_dispersion.py
import matplotlib.pyplot as plt
import numpy as np

from mumax5 import Ferromagnet, Grid, World
from mumax5.util import show_field

# Ferromagnetic spinwave dispersion relation

# Disregarding dipole-dipole interactions, the dispersion relation f(k) in a ferromagnetic wire is given by

# f(k) = (gamma / 2 * pi) * ((2A / M_s) * k^2 + B)

# with A the exchange stiffness, M_s the saturation magnetization, and B an externally applied field.
# We will try to reproduce this dispersion relation numerically using mumax.

# The mumax script below simulates a uniformly magnetized nanowire with an applied field perpendicular to the wire.
# Spin waves are excited by a sinc pulse with a maximum frequency of 20GHz at the center of the simulation box.
# Note that the demagnetization is disabled in this simulation.

# Numerical parameters
fmax = 20E9          # maximum frequency (in Hz) of the sinc pulse
T = 1E-8             # simulation time (longer -> better frequency resolution)
dt = 1 / (2 * fmax)  # the sample time
dx = 4E-9            # cellsize
nx = 1024            # number of cells

# Material/system parameters
Bz = 0.2           # bias field along the z direction
A = 13E-12         # exchange constant
Ms = 800E3         # saturation magnetization
alpha = 0.05       # damping parameter
gamma = 1.76E11    # gyromagnetic ratio

# Create the world
grid_size = (nx, 1, 1)
cell_size = (dx, dx, dx)

world = World(cell_size)
world.bias_magnetic_field = (0, 0, Bz)

# Create a ferromagnet
magnet = Ferromagnet(world, Grid(size=grid_size))
magnet.enable_demag = False
magnet.msat = Ms
magnet.aex = A
magnet.alpha = alpha

Bt = lambda t: (0.01 * np.sinc(2 * fmax * (t - T / 2)), 0, 0)
mask = np.zeros(shape=(1, 1, nx))
# Put signal at the center of the simulation box
mask[:, :, 511:513] = 1
magnet.bias_magnetic_field.add_time_term(Bt, mask)

magnet.magnetization = (0, 0, 1)
magnet.minimize()

timepoints = np.linspace(0, T, 1 + int(T / dt))
outputquantities = {
    'm': lambda: magnet.magnetization.eval(),
    'mx_mean': lambda: magnet.magnetization.average()[0],
    'my_mean': lambda: magnet.magnetization.average()[1],
    'mz_mean': lambda: magnet.magnetization.average()[2],
}

# Run solver
output = world.timesolver.solve(timepoints, outputquantities)

# Plot output data
t = output['time']
mx_mean = output['mx_mean']
my_mean = output['my_mean']
mz_mean = output['mz_mean']

fig, axs = plt.subplots(nrows=3, ncols=1, sharex=True, figsize=(10, 8))
axs[0].plot(t, mx_mean)
axs[0].set_ylabel('$M_x$')

axs[1].plot(t, my_mean)
axs[1].set_ylabel('$M_y$')

axs[2].plot(t, mz_mean)
axs[2].set_ylabel('$M_z$')
axs[2].set_xlabel('$t$')

plt.suptitle('Ferromagnetic Spinwave Dispersion Relation (Python)')
plt.show()

# Apply the two dimensional FFT
m = np.array(output['m'], dtype=float)
mx = m[:, 0, 0, 0, :]
mx_fft = np.fft.fft2(mx)
mx_fft = np.fft.fftshift(mx_fft)

plt.figure(figsize=(10, 6))

# Show the intensity plot of the 2D FFT
extent = [-(2 * np.pi) / (2 * dx), (2 * np.pi) / (2 * dx), -1 /
          (2 * dt), 1 / (2 * dt)]  # extent of k values and frequencies
plt.imshow(np.abs(mx_fft)**2, extent=extent,
           aspect='auto', origin='lower', cmap="inferno")

# Plot the analytical derived dispersion relation
k = np.linspace(-2e8, 2e8, 1000)
freq_theory = A * gamma * k**2 / (np.pi * Ms) + gamma * Bz / (2 * np.pi)
plt.plot(k, freq_theory, 'r--', lw=1)
plt.axhline(gamma * Bz / (2 * np.pi), c='g', ls='--', lw=1)

plt.xlim([-2e8, 2e8])
plt.ylim([0, fmax])
plt.ylabel("$f$ (Hz)")
plt.xlabel("$k$ (1/m)")
plt.title('Ferromagnetic Spinwave Dispersion Relation (Python)')

plt.show()

File: src/cmd/spinwave_dispersion.cpp
#include <cmath>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <limits>
#include <vector>

#include "dynamicequation.hpp"
#include "examples.hpp"
#include "ferromagnet.hpp"
#include "field.hpp"
#include "grid.hpp"
#include "mumaxworld.hpp"
#include "timesolver.hpp"
#include "torque.hpp"

void spinwave_dispersion() {
  std::cout << "*********** Spinwave Dispersion ***********" << std::endl;

  // NUMERICAL PARAMETERS
  real fmax = 20E9;  // maximum frequency(in Hz) of the sinc pulse
  real T = 1E-8;     // simulation time(longer->better frequency resolution)
  real dt = 1 / (2.0 * fmax);  // the sample time
  real dx = 4E-9;              // cellsize
  int nx = 1024;               // number of cells

  // MATERIAL / SYSTEM PARAMETERS
  real Bz = 0.2;         // Bias field along the z direction
  real Aex = 13E-12;     // exchange constant
  real Ms = 800E3;       // saturation magnetization
  real alpha = 0.05;     // damping parameter
  real gamma = 1.76E11;  // gyromagnetic ratio

  int3 grid_size{nx, 1, 1};
  real3 cellsize{dx, dx, dx};
  std::string ferromagnet_name = "";

  MumaxWorld mWorld(cellsize);
  Grid mGrid(grid_size);
  auto magnet = mWorld.addFerromagnet(mGrid, 3, ferromagnet_name);

  magnet->enableDemag = false;
  magnet->msat.set(Ms);
  magnet->aex.set(Aex);
  magnet->alpha.set(alpha);
  magnet->magnetization()->set(real3{0, 0, 1});

  real3 B_ext{0, 0, Bz};
  mWorld.biasMagneticField = B_ext;

  // rectangular region, 2x1, defregion(1, rect(2 * {dx}, inf))
  int ncomp = 3;
  Field mask(magnet->system(), ncomp);
  // set the field to zero outside the given boundary
  std::vector<real> maskValues(nx * ncomp, 0);
  int nCellsDyn = 2;
  // the sinc signal will be located at the center of the simulation box
  for (int i = 0, start_pos = (nx / 2) - 1; i < ncomp; i++) {
    maskValues[i * nx + start_pos] = 1;
    maskValues[i * nx + start_pos + 1] = 1;
  }
  mask.setData(maskValues);

  auto sinc = [](real x) -> real {
    if (std::abs(x) <= std::numeric_limits<real>::epsilon())
      return 1.0;
    else
      return sin(x) / x;
  };

  auto Bt = [&sinc, fmax, T](real t) {
    return real3{0.01F * sinc(2 * M_PI * fmax * (t - T / 2)), 0, 0};
  };

  magnet->biasMagneticField.addTimeDependentTerm(Bt);

  magnet->minimize();

  // --- SCHEDULE THE OUTPUT ---
  real start = 0, stop = T;
  int n_timepoints = static_cast<int>(1 + (stop - start) / dt);

  std::string out_file_path = "spinwave_dispersion_m.csv";
  std::ofstream magn_csv(out_file_path);
  magn_csv.precision(10);
  magn_csv << "t,mx,my,mz,bx" << std::endl;

  std::cout << "Running simulation..." << std::endl;

  for (int i = 0; i < n_timepoints; i++) {
    mWorld.timesolver().run(dt);
    auto m = magnet->magnetization()->average();
    magn_csv << mWorld.time() << "," << m[0] << "," << m[1] << "," << m[2]
             << "," << Bt(i * dt).x << "," << std::endl;
  }

  std::cout << "Simulation results were saved into\n"
            << std::filesystem::current_path().append(out_file_path).string()
            << std::endl;
}
